core/src/main/java/discord4j/core/object/Template.java

```diff
@@ -0,0 +1,186 @@
+package discord4j.core.object;
+
+import discord4j.common.util.Snowflake;
+import discord4j.core.GatewayDiscordClient;
+import discord4j.core.object.entity.Guild;
+import discord4j.core.object.entity.User;
+import discord4j.core.spec.GuildTemplateEditSpec;
+import discord4j.core.spec.TemplateCreateGuildSpec;
+import discord4j.discordjson.json.SerializedSourceGuildData;
+import discord4j.discordjson.json.TemplateData;
+import reactor.core.publisher.Mono;
+
+import java.util.Objects;
+import java.util.function.Consumer;
+
+public class Template implements DiscordObject {
+
+    /** The gateway associated to this object. */
+    private final GatewayDiscordClient gateway;
+
+    /** The raw data as represented by Discord. */
+    private final TemplateData data;
+
+    /** The ID of the guild this template is associated to. */
+    private final long guildId;
+
+    /**
+     * Constructs a {@code Template} with an associated ServiceMediator and Discord data.
+     *
+     * @param gateway The {@link GatewayDiscordClient} associated to this object, must be non-null.
+     * @param data The raw data as represented by Discord, must be non-null.
+     */
+    public Template(final GatewayDiscordClient gateway, final TemplateData data) {
+        this.gateway = Objects.requireNonNull(gateway);
+        this.data = Objects.requireNonNull(data);
+        this.guildId = Long.parseLong(data.sourceGuildId());
+    }
+
+    @Override
+    public GatewayDiscordClient getClient() {
+        return this.gateway;
+    }
+
+    /**
+     * Gets the template code (unique ID).
+     *
+     * @return The template code (unique ID).
+     */
+    public final String getCode() {
+        return data.code();
+    }
+
+    /**
+     * Gets the source guild ID.
+     *
+     * @return The guild id.
+     */
+    public final long getGuildId() {
+        return guildId;
+    }
+
+    /**
+     * Gets the name of the template.
+     *
+     * @return The template name.
+     */
+    public final String getName() {
+        return data.name();
+    }
+
+    /**
+     * Gets the description of the template.
+     *
+     * @return The template description.
+     */
+    public final String getDescription() {
+        return data.description().orElse("");
+    }
```

Comment:
Any reason this is String instead of Optional? We try to stick to Discord original types as much as possible

Reply:
I don't usually use Optionals so that was just habit.